pkg/probe/probe.go

```diff
@@ -0,0 +1,50 @@
+// (C) Copyright Confidential Containers Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package probe
+
+import (
+	"log"
+	"net/http"
+	"os"
+)
+
+var logger = log.New(log.Writer(), "[probe/probe] ", log.LstdFlags|log.Lmsgprefix)
+var readizProbesDone bool
+var checker Checker
+
+const DEFAULT_CC_RUNTIMECLASS_NAME string = "kata-remote"
+
+func StartupHandler(w http.ResponseWriter, r *http.Request) {
+	if !readizProbesDone {
+		ret, err := checker.AllDedicatedPeerPodsAreReady()
+		readizProbesDone = ret
+		if err != nil || !readizProbesDone {
+			logger.Printf("Not all PeerPods ready, because %s", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+	logger.Printf("All PeerPods standup. we do not check the PeerPods status any more.")
+	w.WriteHeader(http.StatusOK)
+}
+
+func Start() {
```

Comment:
So the startup probe is checking for peer-pods in ready state which is good.
One additional scenario that I see missing is that the probe can be ready but the CAA errors out due to some bad config or error. Imho, the probe readiness should also include checking if CAA being ready in addition to the peer-pods in ready state.
Wdyt ?

Reply:
Yes, agreed, let me think about the conditions to be checked and add it. @bpradipt

Reply:
@yoheiueda what's your opinion? What should be checked if we want to add probe to check the services readiness?

Reply:
@bpradipt @yoheiueda do you think the caa unix domain socket a valid testify method in readiness probe?

Reply:
Added uds check in startup probe.

Reply:
@bpradipt caa pod will be recreated according to restart policy if startup probe failed eventually.

Reply:
```
A startup probe is used to determine whether a pod has successfully started. If the startup probe fails, it indicates that the container within the pod has not successfully initialized. Kubernetes considers the pod's startup as unsuccessful if the startup probe fails. The system will then follow the configured actions based on the restartPolicy of the pod and other relevant factors.

Restart Behavior:
The pod's restart behavior depends on the restartPolicy defined in its configuration:

    Always: If the restartPolicy is set to Always, Kubernetes will attempt to restart the pod to bring it to a healthy state. If the startup probe continues to fail after several restart attempts, Kubernetes might give up and mark the pod as "CrashLoopBackOff."
    OnFailure: If the restartPolicy is set to OnFailure, Kubernetes will restart the pod only if it terminates with an error. Failed startup probes may lead to pod restarts in this case.
```